crates/ra_ide_api_light/src/assists/introduce_variable.rs

```diff
@@ -26,6 +26,11 @@ pub fn introduce_variable<'a>(ctx: AssistCtx) -> Option<Assist> {
             false
         };
         if is_full_stmt {
+            if let Some(last_child) = expr.syntax().last_child() {
+                if last_child.kind() != SEMI && !is_semi_right_after(expr.syntax()) {
```

Comment:
@matklad I don't know how to correctly add `pub fn has_semi(&self) -> bool` to `Expr`, since `Expr` seems to be autogenerated.

Reply:
adding 

```
impl ExprStmt {
    pub fn has_semi(&self) -> bool {
        match self.syntax().last_child() {
            None => false,
            Some(node) => node.kind() == SEMI,
        }
    }
}
```

to ast.rs should work I think. Only `generated.rs` is genrated. We actually should generate those `has_X` as well, but that's not pressing

Reply:
This is not exactly what I had in mind. We should use `expr_stmt` from `if let Some(expr_stmt) = ast::ExprStmt::cast(anchor_stmt)` here. That is, we need to save that `expr_stmt` into a variable.